### PR TITLE
Stringatom

### PR DESCRIPTION
**Pull request by daschuer**
This is a work in progress PR for the new StringAtom class. 
It aims to be a 1:1 replacement of all programmatic strings. 

It will finally fix:
https://bugs.launchpad.net/mixxx/+bug/1283471

As discussed it Launchpad, it is desirable to have a index based lookup instead of a pointer lookup, which is implemented here. It turns out the we have too many and dynamic growing groups string amount to have an effective index lookup. 
And we have also at least two independent domains for such indexes: the Master Channels (currently limited by 32) and the Effect Channels (dynamic) 
I plan to introduce an inherited class, that has an additional index per indexing domain, assigned be EffectManager and EngineMaster.  
I hope the code stays maintainable but becomes faster. 

Any Ideas or comments?

### CODE REVIEW

**Comment by daschuer** [comment]
This is intended to be part of 1.12 since this it should have a nice performance boost. 
I am curious, if I am able to proof this.

File: src/control/stringatom.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        : m_pString(pString) {
+}
+
+StringAtom::~StringAtom() {
+    // do not delete m_pString here, it is owned by the s_stringHash
+}
+
+void StringAtom::initalize(const QString& string) {
+    QMutexLocker locker(&s_stringHashMutex);
+    const QString* pString = getInner(string);
```
Comment (rryan) [inline src/control/stringatom.cpp]: You can avoid the double-hash lookup on insert by using ```QHash::operator[]``` which returns a mutable reference to the position in the hash.

```c++
QString* &pString = s_stringHashMutex[string];
if (pString == NULL) {
  pString = new QString(string);
}
```

File: src/control/stringatom.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+/*
+StringAtom& StringAtom::operator= (const QString & string) {
+    return *this;
+}
+*/
+
+QByteArray StringAtom::toAscii() const {
+    return m_pString->toAscii();
```
Comment (rryan) [inline src/control/stringatom.cpp]: m_pString is NULL if default constructor is used

**Comment by rryan** [comment]
Hey Daniel,

It looks like you'd like to replace all group strings across Mixxx with atoms. That's possibly useful but I don't have data to back that up. Whereas I do have data to back up the fact that QHash use in the engine is probably a bad thing. 

Looking at your proposed changes to PlayerManager, I don't think I see the value added here. It looks like it actually adds work to every call. Beyond this, it makes every call to PlayerManager::getGroupForXXX method lock a static shared-state mutex! That's a step backwards as it adds priority inversion to any thread that uses those helper methods now.

Where I think this idea can add value is specifically in the engine. I think the engine should keep track of every group that has been created and hand out "handles" for those groups. Instead of passing strings, we pass handles. If handles are defined as starting at 0 and growing upwards then we can use a QVarLengthArray or QVector to do associative lookups instead of QHash. QHash lookups are the single biggest waste of CPU cycle in the engine right now! This would be hugely useful to replace those with memory lookups (base pointer + handle). 

Also, I'm in general against hidden shared state usage -- especially for things that act like values rather than objects with state. If the compiler coerces a QString to a StringAtom then you have no idea that it just locked a static mutex and did a hash lookup. I think it needs to be quite verbose so that the developer knows what she is doing. ControlObject::getControl(...) sets off warning lights to me for this reason.

Contrived example:
```c++
EffectProcessor::process(const QString& group, CSAMPLE* pInOut, int length) {
  blah(m_groupState.getState(group), pInOut, length); 
}
```
Assuming groupState is a QHash of StringAtom, you don't see the hidden cost of converting the string to a StringAtom. I think it should be loud and clear, like: ```Handle IncrementingIntegerHandle::getOrCreateHandle(groupString)```

> It turns out the we have too many and dynamic growing groups string amount to have an effective index lookup. 

I don't think I understand -- what's "too many"? We don't have many group strings.

**Comment by rryan** [comment]
> It turns out the we have too many and dynamic growing groups string amount to have an effective index lookup.

Like, if we had 1000 groups, that's 1000 pointers. Let's say they're all 16 bytes long. 24kB (on 64-bit) isn't too bad.

File: src/playermanager.cpp [hunk 3]
```diff
@@ -473,3 +471,54 @@ void PlayerManager::slotLoadTrackIntoNextAvailableSampler(TrackPointer pTrack) {
...
+        return group;
+    }
+    DEBUG_ASSERT_AND_HANDLE(i < groupForSamplerList.count()) {
+        return StringAtom();
+    }
+    return groupForSamplerList.at(i);
+}
+
+//static
+StringAtom PlayerManager::groupForDeck(int i, bool add) {
```
Comment (rryan) [inline src/playermanager.cpp]: This isn't thread safe.

File: src/control/stringatom.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+#include <QHash>
+#include <QMutex>
+#include <QString>
+#include <QAtomicPointer>
+#include <QMetaType>
+
+class StringAtom {
+  public:
+    StringAtom();
+    StringAtom(const char* asciiString);
```
Comment (rryan) [inline src/control/stringatom.h]: these constructors should be marked explicit so conversions are noticeable

**Comment by rryan** [comment]
Hi Daniel,

I just reviewed your latest commits and I think you should stop work on this branch until you can answer the following:

1) How will we evaluate that this is a positive change for Mixxx?

2) How will you address the thread priority inversion issues caused by StringAtom creation? Every time you create a StringAtom from a string literal or a QString it involves locking a static mutex. This could have unintended consequences as you create StringAtoms in previously mutex-free helper methods like PlayerManager::groupForXXX. 

3) What is the benefit of this branch over something like this:

```c++
class StringAtom {
  explicit StringAtom(const QString& string)
      : m_string(string), 
        m_hash(qHash(string)) {
  }

  inline bool operator==(const StringAtom& other) const {
    return m_hash == other.m_hash && m_string == other.m_string;
  } 

  private:
    QString m_string;
    uint m_hash; 
}; 

inline uint qHash(const StringAtom& atom) {
  return atom.m_hash;
}
```

The benefits of this are:
* No thread priority issues.
* No time spent hashing strings when used as a QHash key.
* Lightweight-ish (integer) comparisons unless there is a hash collision or string equality.

The downsides are:
* Small cost in pre-hashing the string when constructing from a QString or string literal.
* For ```StringAtom::operator==``` in the equality case ```QString::operator==``` must be used, resulting in roughly ~12 character comparisons (our group names are typically short).

The benefits of the current StringAtom in your PR:
* No time spent hashing strings when used as a QHash key.
* Truly lightweight (pointer) comparisons.

The downsides of the current StringAtom in your PR:
* Lock a global mutex when constructing from a QString or string literal.

**Comment by rryan** [comment]
Also, none of the changes you've currently made touch any of the known hotspots in the engine. I don't see anything that would give a performance boost (which you were saying was why you wanted this in 1.12).

**Comment by rryan** [comment]
> Where I think this idea can add value is specifically in the engine. I think the engine should keep track of every group that has been created and hand out "handles" for those groups.

I have a PR coming that makes this change so no need to expand this PR to do that.

**Comment by Be-ing** [comment]
This pull request has not been updated in years so I am closing it to reduce the clutter of open PRs. It also does not seem to be relevant anymore now that we have ChannelHandles.

**Pull request closed without merging**